File: src/harmonic/audioAxis.ts
/**
 * @file audioAxis.ts
 * @module harmonic/audioAxis
 * @layer Layer 14
 * @component Audio Telemetry Engine
 * @version 3.0.0
 * @since 2026-01-20
 *
 * SCBE Audio Axis - FFT-based telemetry without altering the invariant metric.
 *
 * Layer 14: f_audio(t) = [E_a, C_a, F_a, r_HF,a]
 *
 * Features:
 * - E_a = log(ε + Σₙ a[n]²) — Frame energy
 * - C_a = (Σₖ fₖ·Pₐ[k]) / (Σₖ Pₐ[k]) — Spectral centroid
 * - F_a = Σₖ (√Pₐ[k] - √Pₐ_prev[k])² — Spectral flux
 * - r_HF,a = Σₖ∈Khigh Pₐ[k] / Σₖ Pₐ[k] — High-frequency ratio
 * - S_audio = 1 - r_HF,a — Audio stability score ∈ [0,1]
 */

/** Small epsilon for numerical stability */
const EPSILON = 1e-10;

/**
 * Audio features extracted from a frame
 */
export interface AudioFeatures {
  /** Frame energy: Ea = log(ε + Σₙ a[n]²) */
  energy: number;
  /** Spectral centroid: Ca = (Σₖ fₖ·Pₐ[k]) / (Σₖ Pₐ[k]) */
  centroid: number;
  /** Spectral flux: Fa = Σₖ (√Pₐ[k] - √Pₐ_prev[k])² */
  flux: number;
  /** High-frequency ratio: rHF,a = Σₖ∈Khigh Pₐ[k] / Σₖ Pₐ[k] */
  hfRatio: number;
  /** Audio stability score: Saudio = 1 - rHF,a */
  stability: number;
}

/**
 * Audio Axis Processor configuration
 */
export interface AudioAxisConfig {
  /** Sample rate in Hz (default: 44100) */
  sampleRate?: number;
  /** FFT size (default: 2048) */
  fftSize?: number;
  /** High frequency cutoff ratio (default: 0.5 = half Nyquist) */
  hfCutoff?: number;
  /** Risk weight for audio channel (default: 0.1) */
  riskWeight?: number;
}

/**
 * Simple DFT implementation (for environments without Web Audio API)
 * Returns magnitude spectrum (power)
 */
function computeDFT(signal: number[]): number[] {
  const N = signal.length;
  const spectrum: number[] = new Array(Math.floor(N / 2) + 1);

  for (let k = 0; k < spectrum.length; k++) {
    let re = 0,
      im = 0;
    for (let n = 0; n < N; n++) {
      const angle = (2 * Math.PI * k * n) / N;
      re += signal[n] * Math.cos(angle);
      im -= signal[n] * Math.sin(angle);
    }
    // Power spectrum: |X[k]|²
    spectrum[k] = (re * re + im * im) / N;
  }

  return spectrum;
}

/**
 * Audio Axis Processor - Layer 14 telemetry
 */
export class AudioAxisProcessor {
  private sampleRate: number;
  private fftSize: number;
  private hfCutoff: number;
  private riskWeight: number;
  private prevSpectrum: number[] | null = null;

  constructor(config: AudioAxisConfig = {}) {
    this.sampleRate = config.sampleRate ?? 44100;
    this.fftSize = config.fftSize ?? 2048;
    this.hfCutoff = config.hfCutoff ?? 0.5;
    this.riskWeight = config.riskWeight ?? 0.1;
  }

  /**
   * Compute frame energy
   * Ea = log(ε + Σₙ a[n]²)
   */
  private computeEnergy(signal: number[]): number {
    let sum = 0;
    for (const s of signal) {
      sum += s * s;
    }
    return Math.log(EPSILON + sum);
  }

  /**
   * Compute spectral centroid
   * Ca = (Σₖ fₖ·Pₐ[k]) / (Σₖ Pₐ[k])
   */
  private computeCentroid(spectrum: number[]): number {
    let weightedSum = 0;
    let totalPower = 0;
    const binWidth = this.sampleRate / this.fftSize;

    for (let k = 0; k < spectrum.length; k++) {
      const freq = k * binWidth;
      weightedSum += freq * spectrum[k];
      totalPower += spectrum[k];
    }

    return weightedSum / (totalPower + EPSILON);
  }

  /**
   * Compute spectral flux
   * Fa = Σₖ (√Pₐ[k] - √Pₐ_prev[k])²
   */
  private computeFlux(spectrum: number[], prevSpectrum: number[] | null): number {
    if (!prevSpectrum) return 0;

    let flux = 0;
    const len = Math.min(spectrum.length, prevSpectrum.length);

    for (let k = 0; k < len; k++) {
      const diff = Math.sqrt(spectrum[k]) - Math.sqrt(prevSpectrum[k]);
      flux += diff * diff;
    }

    // Normalize by number of bins
    return flux / len;
  }

  /**
   * Compute high-frequency ratio
   * rHF,a = Σₖ∈Khigh Pₐ[k] / Σₖ Pₐ[k]
   */
  private computeHFRatio(spectrum: number[]): number {
    const cutoffBin = Math.floor(spectrum.length * this.hfCutoff);
    let hfPower = 0;
    let totalPower = 0;

    for (let k = 0; k < spectrum.length; k++) {
      totalPower += spectrum[k];
      if (k >= cutoffBin) {
        hfPower += spectrum[k];
      }
    }

    return hfPower / (totalPower + EPSILON);
  }

  /**
   * Process an audio frame and extract features
   *
   * @param signal - Audio samples (time domain)
   * @returns Audio features
   */
  processFrame(signal: number[]): AudioFeatures {
    // Compute spectrum via DFT
    const spectrum = computeDFT(signal);

    // Extract features
    const energy = this.computeEnergy(signal);
    const centroid = this.computeCentroid(spectrum);
    const flux = this.computeFlux(spectrum, this.prevSpectrum);
    const hfRatio = this.computeHFRatio(spectrum);
    const stability = 1 - hfRatio;

    // Store for next frame
    this.prevSpectrum = spectrum;

    return { energy, centroid, flux, hfRatio, stability };
  }

  /**
   * Integrate audio risk into base risk
   * Risk' = Risk_base + wa·(1 - Saudio)
   *
   * @param baseRisk - Base risk value
   * @param features - Audio features
   * @returns Adjusted risk
   */
  integrateRisk(baseRisk: number, features: AudioFeatures): number {
    return baseRisk + this.riskWeight * (1 - features.stability);
  }

  /**
   * Reset processor state
   */
  reset(): void {
    this.prevSpectrum = null;
  }

  /**
   * Get configuration
   */
  getConfig(): AudioAxisConfig {
    return {
      sampleRate: this.sampleRate,
      fftSize: this.fftSize,
      hfCutoff: this.hfCutoff,
      riskWeight: this.riskWeight,
    };
  }
}

/**
 * Generate a test signal (sine wave)
 */
export function generateTestSignal(
  frequency: number,
  duration: number,
  sampleRate: number = 44100
): number[] {
  const samples = Math.floor(duration * sampleRate);
  const signal: number[] = new Array(samples);

  for (let n = 0; n < samples; n++) {
    signal[n] = Math.sin((2 * Math.PI * frequency * n) / sampleRate);
  }

  return signal;
}

/**
 * Generate white noise
 */
export function generateNoise(samples: number): number[] {
  const signal: number[] = new Array(samples);
  for (let n = 0; n < samples; n++) {
    signal[n] = Math.random() * 2 - 1;
  }
  return signal;
}

File: SCBE-AETHERMOORE-v3.0.0/src/symphonic_cipher/scbe_aethermoore/axiom_grouped/audio_axis.py
#!/usr/bin/env python3
"""
Layer 14: Audio Axis - FFT-based Telemetry Channel

Audio telemetry provides deterministic features for alerting, drift detection,
and risk scoring without altering the invariant hyperbolic metric.

Feature Extraction (from FFT/STFT):
- Ea = log(ε + Σn a[n]²)             [Frame energy, log-scale]
- Ca = (Σk fk·Pa[k]) / (Σk Pa[k])   [Spectral centroid]
- Fa = Σk (√Pa[k] - √Pa_prev[k])²   [Spectral flux]
- rHF,a = Σk∈Khigh Pa[k] / Σk Pa[k] [High-frequency ratio]
- Saudio = 1 - rHF,a                [Audio stability score]

Risk Integration:
Risk' = Risk_base + wa·(1 - Saudio)
      = Risk_base + wa·rHF,a

Document ID: SCBE-L14-2026-001
Version: 1.0.0
"""

import math
import numpy as np
from dataclasses import dataclass, field
from typing import List, Tuple, Optional, Dict, Any
from enum import Enum


# =============================================================================
# CONSTANTS
# =============================================================================

# Default FFT parameters
DEFAULT_N_FFT = 256
DEFAULT_HF_FRAC = 0.3  # Top 30% = high frequency
DEFAULT_EPSILON = 1e-10
DEFAULT_SAMPLE_RATE = 16000  # 16 kHz

# Audio weight in risk formula
DEFAULT_WA = 0.2


class AudioStability(Enum):
    """Audio stability classification."""
    STABLE = "STABLE"           # Saudio >= 0.7
    MODERATE = "MODERATE"       # 0.4 <= Saudio < 0.7
    UNSTABLE = "UNSTABLE"       # 0.2 <= Saudio < 0.4
    CRITICAL = "CRITICAL"       # Saudio < 0.2


# =============================================================================
# AUDIO FEATURE EXTRACTION
# =============================================================================

@dataclass
class AudioFeatures:
    """
    Layer 14 audio telemetry features.
    
    faudio(t) = [Ea, Ca, Fa, rHF,a]
    Saudio = 1 - rHF,a
    """
    energy: float           # Ea: log frame energy
    centroid: float         # Ca: spectral centroid (Hz)
    flux: float             # Fa: spectral flux
    hf_ratio: float         # rHF,a: high-frequency ratio
    stability: float        # Saudio = 1 - rHF,a
    classification: AudioStability
    timestamp: float = 0.0
    
    def to_vector(self) -> np.ndarray:
        """Return feature vector [Ea, Ca, Fa, rHF,a]."""
        return np.array([self.energy, self.centroid, self.flux, self.hf_ratio])
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert to dictionary."""
        return {
            "energy": self.energy,
            "centroid": self.centroid,
            "flux": self.flux,
            "hf_ratio": self.hf_ratio,
            "stability": self.stability,
            "classification": self.classification.value,
            "timestamp": self.timestamp
        }


def compute_fft(signal: np.ndarray, n_fft: int = DEFAULT_N_FFT) -> Tuple[np.ndarray, np.ndarray]:
    """
    Compute FFT of audio frame.
    
    A[k] = Σn a[n]·e^(-i2πkn/N)
    Pa[k] = |A[k]|²
    
    Args:
        signal: Audio samples
        n_fft: FFT size
        
    Returns:
        (frequencies, power_spectrum)
    """
    if len(signal) < n_fft:
        signal = np.pad(signal, (0, n_fft - len(signal)))
    elif len(signal) > n_fft:
        signal = signal[:n_fft]
    
    # Apply Hann window
    window = np.hanning(n_fft)
    windowed = signal * window
    
    # FFT
    fft_result = np.fft.rfft(windowed)
    power = np.abs(fft_result) ** 2
    
    # Frequency bins
    freqs = np.fft.rfftfreq(n_fft, 1.0 / DEFAULT_SAMPLE_RATE)
    
    return freqs, power


def extract_energy(signal: np.ndarray, eps: float = DEFAULT_EPSILON) -> float:
    """
    Extract frame energy (log-scale).
    
    Ea = log(ε + Σn a[n]²)
    
    Args:
        signal: Audio samples
        eps: Small constant for numerical stability
        
    Returns:
        Log energy
    """
    energy = np.sum(signal ** 2)
    return float(np.log(eps + energy))


def extract_centroid(freqs: np.ndarray, power: np.ndarray, 
                     eps: float = DEFAULT_EPSILON) -> float:
    """
    Extract spectral centroid.
    
    Ca = (Σk fk·Pa[k]) / (Σk Pa[k] + ε)
    
    Args:
        freqs: Frequency bins
        power: Power spectrum
        eps: Small constant
        
    Returns:
        Spectral centroid in Hz
    """
    total_power = np.sum(power) + eps
    centroid = np.sum(freqs * power) / total_power
    return float(centroid)


def extract_flux(power: np.ndarray, prev_power: Optional[np.ndarray],
                 eps: float = DEFAULT_EPSILON) -> float:
    """
    Extract spectral flux.
    
    Fa = Σk (√Pa[k] - √Pa_prev[k])² / (Σk Pa[k] + ε)
    
    Args:
        power: Current power spectrum
        prev_power: Previous power spectrum (None for first frame)
        eps: Small constant
        
    Returns:
        Spectral flux
    """
    if prev_power is None:
        return 0.0
    
    sqrt_power = np.sqrt(power)
    sqrt_prev = np.sqrt(prev_power)
    
    diff_sq = (sqrt_power - sqrt_prev) ** 2
    total_power = np.sum(power) + eps
    
    flux = np.sum(diff_sq) / total_power
    return float(flux)


def extract_hf_ratio(power: np.ndarray, hf_frac: float = DEFAULT_HF_FRAC,
                     eps: float = DEFAULT_EPSILON) -> float:
    """
    Extract high-frequency ratio.
    
    rHF,a = Σk∈Khigh Pa[k] / (Σk Pa[k] + ε)
    
    where Khigh is the top hf_frac of frequency bins.
    
    Args:
        power: Power spectrum
        hf_frac: Fraction of bins considered "high frequency"
        eps: Small constant
        
    Returns:
        High-frequency ratio in [0, 1]
    """
    n_bins = len(power)
    hf_start = int(n_bins * (1 - hf_frac))
    
    hf_power = np.sum(power[hf_start:])
    total_power = np.sum(power) + eps
    
    ratio = hf_power / total_power
    return float(np.clip(ratio, 0.0, 1.0))


def classify_stability(saudio: float) -> AudioStability:
    """
    Classify audio stability.
    
    Args:
        saudio: Audio stability score (1 - rHF,a)
        
    Returns:
        AudioStability classification
    """
    if saudio >= 0.7:
        return AudioStability.STABLE
    elif saudio >= 0.4:
        return AudioStability.MODERATE
    elif saudio >= 0.2:
        return AudioStability.UNSTABLE
    else:
        return AudioStability.CRITICAL


def extract_audio_features(
    signal: np.ndarray,
    prev_power: Optional[np.ndarray] = None,
    n_fft: int = DEFAULT_N_FFT,
    hf_frac: float = DEFAULT_HF_FRAC,
    timestamp: float = 0.0
) -> Tuple[AudioFeatures, np.ndarray]:
    """
    Extract all Layer 14 audio features from a frame.
    
    Args:
        signal: Audio samples
        prev_power: Previous frame's power spectrum
        n_fft: FFT size
        hf_frac: High-frequency fraction
        timestamp: Frame timestamp
        
    Returns:
        (AudioFeatures, current_power) for chaining
    """
    # FFT
    freqs, power = compute_fft(signal, n_fft)
    
    # Extract features
    energy = extract_energy(signal)
    centroid = extract_centroid(freqs, power)
    flux = extract_flux(power, prev_power)
    hf_ratio = extract_hf_ratio(power, hf_frac)
    
    # Stability score
    stability = 1.0 - hf_ratio
    classification = classify_stability(stability)
    
    features = AudioFeatures(
        energy=energy,
        centroid=centroid,
        flux=flux,
        hf_ratio=hf_ratio,
        stability=stability,
        classification=classification,
        timestamp=timestamp
    )
    
    return features, power


# =============================================================================
# RISK INTEGRATION
# =============================================================================

def audio_risk_additive(
    risk_base: float,
    saudio: float,
    wa: float = DEFAULT_WA
) -> float:
    """
    Additive audio risk integration.
    
    Risk' = Risk_base + wa·(1 - Saudio)
          = Risk_base + wa·rHF,a
    
    Args:
        risk_base: Base risk from Layers 1-13
        saudio: Audio stability score
        wa: Audio weight
        
    Returns:
        Adjusted risk
    """
    return risk_base + wa * (1.0 - saudio)


def audio_risk_multiplicative(
    risk_base: float,
    saudio: float,
    wa: float = DEFAULT_WA
) -> float:
    """
    Multiplicative audio risk integration.
    
    Risk' = Risk_base · (1 + wa·rHF,a)
    
    Audio instability amplifies geometric risk.
    
    Args:
        risk_base: Base risk from Layers 1-13
        saudio: Audio stability score
        wa: Audio weight
        
    Returns:
        Adjusted risk
    """
    rHF = 1.0 - saudio
    return risk_base * (1.0 + wa * rHF)


# =============================================================================
# AUDIO AXIS PROCESSOR
# =============================================================================

class AudioAxisProcessor:
    """
    Layer 14 Audio Axis processor.
    
    Maintains state for streaming audio analysis.
    """
    
    def __init__(
        self,
        n_fft: int = DEFAULT_N_FFT,
        hf_frac: float = DEFAULT_HF_FRAC,
        wa: float = DEFAULT_WA,
        integration_mode: str = "additive"
    ):
        """
        Initialize processor.
        
        Args:
            n_fft: FFT size
            hf_frac: High-frequency fraction
            wa: Audio weight for risk integration
            integration_mode: "additive" or "multiplicative"
        """
        self.n_fft = n_fft
        self.hf_frac = hf_frac
        self.wa = wa
        self.integration_mode = integration_mode
        
        # State
        self._prev_power: Optional[np.ndarray] = None
        self._frame_count = 0
        self._history: List[AudioFeatures] = []
        self._max_history = 100
    
    def process_frame(self, signal: np.ndarray) -> AudioFeatures:
        """
        Process a single audio frame.
        
        Args:
            signal: Audio samples
            
        Returns:
            AudioFeatures for this frame
        """
        timestamp = self._frame_count * self.n_fft / DEFAULT_SAMPLE_RATE
        
        features, power = extract_audio_features(
            signal,
            self._prev_power,
            self.n_fft,
            self.hf_frac,
            timestamp
        )
        
        self._prev_power = power
        self._frame_count += 1
        
        # Update history
        self._history.append(features)
        if len(self._history) > self._max_history:
            self._history.pop(0)
        
        return features
    
    def integrate_risk(self, risk_base: float, features: AudioFeatures) -> float:
        """
        Integrate audio features into risk score.
        
        Args:
            risk_base: Base risk from Layers 1-13
            features: Audio features
            
        Returns:
            Adjusted risk
        """
        if self.integration_mode == "multiplicative":
            return audio_risk_multiplicative(risk_base, features.stability, self.wa)
        else:
            return audio_risk_additive(risk_base, features.stability, self.wa)
    
    def get_mean_stability(self) -> float:
        """Get mean stability over history."""
        if not self._history:
            return 1.0
        return float(np.mean([f.stability for f in self._history]))
    
    def get_statistics(self) -> Dict[str, Any]:
        """Get processor statistics."""
        if not self._history:
            return {
                "frame_count": 0,
                "mean_stability": 1.0,
                "mean_energy": 0.0,
                "mean_centroid": 0.0
            }
        
        return {
            "frame_count": self._frame_count,
            "mean_stability": self.get_mean_stability(),
            "mean_energy": float(np.mean([f.energy for f in self._history])),
            "mean_centroid": float(np.mean([f.centroid for f in self._history])),
            "mean_flux": float(np.mean([f.flux for f in self._history])),
            "classifications": {
                s.value: sum(1 for f in self._history if f.classification == s)
                for s in AudioStability
            }
        }
    
    def reset(self):
        """Reset processor state."""
        self._prev_power = None
        self._frame_count = 0
        self._history.clear()


# =============================================================================
# PROOFS AND VERIFICATION
# =============================================================================

def verify_stability_bounded() -> bool:
    """
    Verify: Saudio ∈ [0, 1] for all inputs.
    
    Since Saudio = 1 - rHF,a and rHF,a ∈ [0, 1], Saudio ∈ [0, 1].
    """
    for _ in range(100):
        # Random signal
        signal = np.random.randn(DEFAULT_N_FFT)
        features, _ = extract_audio_features(signal)
        
        if features.stability < 0.0 or features.stability > 1.0:
            return False
        if features.hf_ratio < 0.0 or features.hf_ratio > 1.0:
            return False
    
    return True


def verify_hf_detection() -> bool:
    """
    Verify: High-frequency signals have high rHF,a.
    
    A pure high-frequency sine should have rHF,a close to 1.
    """
    # Low frequency signal (100 Hz)
    t = np.linspace(0, DEFAULT_N_FFT / DEFAULT_SAMPLE_RATE, DEFAULT_N_FFT)
    low_freq = np.sin(2 * np.pi * 100 * t)
    
    # High frequency signal (7000 Hz)
    high_freq = np.sin(2 * np.pi * 7000 * t)
    
    low_features, _ = extract_audio_features(low_freq)
    high_features, _ = extract_audio_features(high_freq)
    
    # High freq should have higher rHF,a
    return high_features.hf_ratio > low_features.hf_ratio


def verify_flux_sensitivity() -> bool:
    """
    Verify: Spectral flux detects changes between frames.
    
    Identical frames should have flux ≈ 0.
    Different frames should have flux > 0.
    """
    signal1 = np.sin(2 * np.pi * 440 * np.linspace(0, 0.016, DEFAULT_N_FFT))
    signal2 = np.sin(2 * np.pi * 880 * np.linspace(0, 0.016, DEFAULT_N_FFT))
    
    # Same signal twice
    _, power1 = extract_audio_features(signal1)
    features_same, _ = extract_audio_features(signal1, power1)
    
    # Different signals
    features_diff, _ = extract_audio_features(signal2, power1)
    
    # Same should have low flux, different should have high flux
    return features_same.flux < features_diff.flux


# =============================================================================
# DEMO
# =============================================================================

if __name__ == "__main__":
    print("=" * 70)
    print("LAYER 14: AUDIO AXIS - FFT-based Telemetry")
    print("=" * 70)
    print()
    
    # Verify proofs
    print("MATHEMATICAL PROOFS:")
    print(f"  Stability bounded [0,1]:  {'✓ PROVEN' if verify_stability_bounded() else '✗ FAILED'}")
    print(f"  HF detection:             {'✓ PROVEN' if verify_hf_detection() else '✗ FAILED'}")
    print(f"  Flux sensitivity:         {'✓ PROVEN' if verify_flux_sensitivity() else '✗ FAILED'}")
    print()
    
    # Demo processing
    print("DEMO PROCESSING:")
    processor = AudioAxisProcessor()
    
    # Simulate stable signal (low frequency)
    t = np.linspace(0, 0.016, DEFAULT_N_FFT)
    stable_signal = np.sin(2 * np.pi * 440 * t)  # 440 Hz
    
    features = processor.process_frame(stable_signal)
    print(f"  Stable signal (440 Hz):")
    print(f"    Energy:     {features.energy:.2f}")
    print(f"    Centroid:   {features.centroid:.0f} Hz")
    print(f"    HF Ratio:   {features.hf_ratio:.3f}")
    print(f"    Stability:  {features.stability:.3f} → {features.classification.value}")
    print()
    
    # Simulate unstable signal (high frequency noise)
    unstable_signal = np.random.randn(DEFAULT_N_FFT)
    features = processor.process_frame(unstable_signal)
    print(f"  Unstable signal (noise):")
    print(f"    Energy:     {features.energy:.2f}")
    print(f"    Centroid:   {features.centroid:.0f} Hz")
    print(f"    HF Ratio:   {features.hf_ratio:.3f}")
    print(f"    Stability:  {features.stability:.3f} → {features.classification.value}")
    print()
    
    # Risk integration demo
    print("RISK INTEGRATION:")
    risk_base = 0.3
    risk_stable = audio_risk_additive(risk_base, 0.8)  # Stable
    risk_unstable = audio_risk_additive(risk_base, 0.2)  # Unstable
    print(f"  Base risk:     {risk_base:.2f}")
    print(f"  + Stable:      {risk_stable:.2f} (Saudio=0.8)")
    print(f"  + Unstable:    {risk_unstable:.2f} (Saudio=0.2)")
    print()
    
    print("=" * 70)
    print("Layer 14 provides deterministic audio telemetry without")
    print("altering the invariant hyperbolic metric dℍ.")
    print("=" * 70)
